### ISSUE TITLE
Revisit linearize transformation

### ISSUE THREAD

**Issue by IgnaceBleukx** [CPMpy/cpmpy#183]
With wsum optimizations (#182) and other changes in place, we should revisit the linearize transformations, as well defining a better "linear normal form" so we know exactly what kind of expressions can come out of the transformations.

Furthermore, we should also define the "supported" parameter as many MIP-solvers support API calls such as `z = AND(x,y)`etc

**Comment by IgnaceBleukx** [comment]
Proposal for a grammer for linear expressions:

```
Linearize returns any of :
1) LinExpr
2) Boolvar -> LinExpr 
3) Expr                       (Expr.name in supported)
-----------------------------------------------
LinExpr:
1) Expr <= Const              (Expr.name in {"sum","wsum"})
2) Expr >= Const              (Expr.name in {"sum","wsum"})
3) Expr == Const              (Expr.name in {"sum","wsum"})
```

This differs from the current specification as variabels can be part of the rhs of LinExpr (except in reified expressions where they have to be constants).
Note the current imlementation is very highly tailored to support as much as possible the API of Gurobi. With Exact, SCIP and potentially other linear solvers in the pipeline, linearization should be generalized somewhat more.

With the above changes in place, `only_var_lhs` and `only_const_rhs` can also go.

**Comment by JoD** [comment]
- I think we also need to add
```
4) BoolVar <- LinExpr
```
- What is meant with `3) Expr   (Expr.name in supported)`? Is this how we allow Gurobi-supported operations (i.e., min/max/abs or multiplication with a Boolean variable)? If so, `max(x,y)` is an `Expr`, but not a constraint (no Boolean type), so I do not see it yet.

**Comment by JoD** [comment]
This grammar is close to what Exact allows: https://gitlab.com/JoD/exact/-/blob/main/src/Exact.hpp. 

In the above notation, Exact also allows:
```
Boolvar <-> LinExpr
Const >= Expr >= Const
```
But it is more restrictive as it only allows nested `LinExpr` of the form
```
Expr >= Const
```
There exist simple conversions between all (splitting `==` or `<->` is a straightforward operation), so I can live with the above grammer.

**Comment by JoD** [comment]
Ok, I'm going to propose two variations: a maximal one (easy for CPMpy) and a minimal one (easy for the solvers). I'm ignoring the "supported" rule for now.
```
**Maximal:**
1) LinExpr
2) Boolvar -> LinExpr
3) Boolvar <- LinExpr
4) Boolvar <-> LinExpr
-----------------------------------------------
LinExpr:
5) Expr <= Const
6) Expr >= Const
7) Expr == Const
-----------------------------------------------
Expr:
8) Sum
9) Wsum
```
```
**Minimal:**
10) LinExpr
11) Boolvar -> LinExpr
12) Boolvar <- LinExpr
-----------------------------------------------
LinExpr:
13) Wsum >= Const
```
What Ignace proposed is pretty close to the maximal one (just drop `4)`). I'm ok with either (simple conversion between both) but it's necessary to consider the variation in options.

**Comment by JoD** [comment]
Also, let's agree that
```
Wsum:
14) (Int IntVar)*
------
Sum:
15) IntVar*
```
So no max/min/abs/mul in there, except when explicitly allowed.

**Comment by IgnaceBleukx** [comment]
Hmm, I am not a fan of the `Boolvar <- LinExpr` variant. We have the `only_bv_implies` transformation which translated any expression `expr -> bv` to `~bv -> ~expr`. I think we can state this has to be done BEFORE linearize is called. I don't think a lot of solvers support `expr -> bv` anyway (Most CP-solvers don't : ) )

Indeed, the `supported` set of operators would be an optional argument to pass to the linearize function so these function can fall trough. We shouldn't consider any difference in reified vs non-refied supported arguments as this should be handled by the `reify_rewrite` transformation BEFORE calling linearize.

14,15 yes indeed. This should be added (but is ensured by flatten_constraint).

**Comment by JoD** [comment]
> Indeed, the supported set of operators would be an optional argument to pass to the linearize function so these function can fall through. We shouldn't consider any difference in reified vs non-refied supported arguments as this should be handled by the reify_rewrite transformation BEFORE calling linearize.

Agree. The remaining question is: how do we let these operators fall through? Is there a difference between keeping some operators and keeping builtin global constraints? What do we allow and what do we not allow? E.g., I guess `max(x,y)+z >= 3` is allowed, but `max(max(x,y),z) >=3` is not.

**Comment by IgnaceBleukx** [comment]
No, there should be no difference between operators or global constraints I think. 
We should build from the Flat Normal Form (FNF) and decide which expressions remain as is and which don't. For your example, both expressions are not flat...
The hard part is to decide which numerical operators to support in combination with what comparisons.... I.e., maybe `max(x,y) = z` is supported but `max(x,y) <= z` is not. The solution is simple: leave it to another transformation, namely `only_numexpr_equality` :) 

Because we only support flat constraints in linearize, any operator name given in the `supported` set should be unambiguous to which part of the constraint it applies.

**Comment by JoD** [comment]
Ok, one issue is that of equality in an equivalence reification. E.g.:
```
p <-> x+y=5
```
is complex to translate, as it entails the following disjunction of two linear inequalities:
```
p | x+y<5 | x+y>5 
```
All other constraints accepted by the grammar translate straightforwardly to a set of linear inequalities _without_ the need for new variables. The above does need new variables. I think the burden of introducing new variables should fall on CPMpy.

**Comment by JoD** [comment]
Wait, the other way around? We do not want 'Lit <-> expr' or 'Lit <- expr' as output for the linearize method, as this is hard to translate for people writing the solver interface? I.e., they need to introduce new variables.

Sidenote: when you write BV, can the variable still be negated?

**Comment by JoD** [comment]
After discussion with Ignace, we're going for the following grammar:
```
Top: 
(Lit ->)? (LinExpr | GenExpr)

Lit:
~? BoolVar // a Boolean variable or its negation

LinExpr: 
(Sum | WSum) (== | >= | <=) Const

Sum:
(+ IntVar)*

WSum:
(+ Const IntVar)*

GenExpr:
Special
Special (== | >= | <=) (Const | IntVar)
```
The reasoning behind `GenExpr` and `Special` is that we want to be able to pass constraints such as `max(x,y) == z` to Gurobi, or selected global constraints to solvers that support them. For other solvers, transformations will be made that linearize these as much as possible.

**Comment by JoD** [comment]
The input grammar accepted by linearize is most of what is output by flatten, as long as all the non-Boolean and non-linear constraints are marked as `Special` (otherwise linearize will throw an error).
